Configurably generate a case id

garethbowen opened the pull request:
# Description

Adds a new configuration option for generating a case identifier.
Show the case_id as a generated field.
Link the case_id field to make it easy to search for related reports.

medic/cht-core#6291

Also solves a bug with the freetext search field not being cleared
when the filters are reset.

medic/cht-core#6356

Also solves a bug where the patient name and patient UUID links
were incorrect.

medic/cht-core#6149

# Code review items

- Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

garethbowen commented on webapp/src/js/actions/reports.js:
```diff
@@ -124,8 +123,9 @@ angular.module('inboxServices').factory('ReportsActions',
-              liveList.setSelected(model.doc && model.doc._id);
+            if (LiveList.reports.initialised()) {
+              LiveList.reports.setSelected(model.doc && model.doc._id);
+              LiveList['report-search'].setSelected(model.doc && model.doc._id);
```
This fixes an unreported bug where when the reports tab was filtered the items never appeared selected.

dianabarsan replied:
Nice catch!

garethbowen commented on webapp/src/js/services/format-data-record.js:
```diff
@@ -55,35 +55,37 @@ angular
+          fields.data.push(result);
           fields.headers.push({ head: titleize(key[0]) });
-          fields.data.push(
-            Object.assign(fieldsToHtml(key[1], labels, data[key[0]], def, locale), {
```
As far as I can tell `fieldToHtml` was being passed entirely the wrong parameters... Now I'm not sure if the `key` is ever an array.

dianabarsan replied:
YES! It can be an array and it seems to still not work properly.
This is the "sample" form I used (that has an Array key):
```
"NESTED": {
        "meta": {
          "code": "NESTED",
          "icon": "",
          "translation_key": "nested"
        },
        "fields": {
          "some.nested.field": {
            "labels": {
              "short": {
                "translation_key": "nested field"
              },
              "tiny": {
                "en": "nested"
              }
            },
            "position": 1,
            "flags": {},
            "type": "string",
            "required": false
          }
        },
        "use_sentinel": true
      },
```

garethbowen replied:
I had a look through and it's not a simple fix. I also assume the feature is not used because it's badly broken. Are you ok with me raising a new issue for this specific use case, so we can either remove support if it's not used, or get it to work properly if it is?

dianabarsan replied:
Definitely ok with raising a separate issue!

dianabarsan commented on webapp/src/templates/partials/reports_content.html:
```diff
@@ -64,14 +64,16 @@
-            <span ng-if="!field.hasUrl">{{field.value}}</span>
+            <span ng-if="field.target && field.target.filter"><a ng-click="reportsContentCtrl.search(field.target.filter)">{{field.value}}</a></span>
+            <span ng-if="field.target && field.target.url"><a ui-sref="{{field.target.url.route}}({{field.target.url.params}})">{{field.value}}</a></span>
+            <span ng-if="!field.target">{{field.value}}</span>
```
I _think_ this also fixes https://github.com/medic/cht-core/issues/6149
3 birds with one stone?

garethbowen replied:
Really? Cool! I'll investigate.

garethbowen replied:
Totally! Added to the PR description.

dianabarsan commented on webapp/src/js/services/format-data-record.js:
```diff
@@ -55,35 +55,37 @@ angular
 
     const fieldsToHtml = function(
       settings,
```
As far as I can tell, this function only runs for SMS reports. 
Even if an XML report had a `case_id` (generated with the transition, so not a `field)`, it wouldn't be displayed. Should this support be added?

garethbowen replied:
I don't think this comment is on the right line, but I see what you mean. This happens with `patient_id` too. I've made a change to include these fields as well, but it's a bit messy because the JSON and XML code is so different. Let me know what you think...

dianabarsan replied:
Awesome, thanks! I think it's good to support generated fields in XML forms as well! 
I also think you cleaned up the template so it's less messy.

dianabarsan commented on webapp/src/templates/partials/reports_content.html:
```diff
@@ -64,14 +64,16 @@
-            <span ng-if="field.hasUrl"><a ui-sref="contacts.detail({ id: reportsContentCtrl.summaries[$index].subject._id || reportsContentCtrl.summaries[$index].fields.patient_uuid })">{{field.value}}</a></span>
-            <span ng-if="!field.hasUrl">{{field.value}}</span>
+            <span ng-if="field.target && field.target.filter"><a ng-click="reportsContentCtrl.search(field.target.filter)">{{field.value}}</a></span>
+            <span ng-if="field.target && field.target.url"><a ui-sref="{{field.target.url.route}}({{field.target.url.params}})">{{field.value}}</a></span>
```
The sad part about this is that when I click on this link, my report gets "deselected". 
https://github.com/medic/cht-core/issues/4825

garethbowen replied:
Yeah, it's a little counter intuitive, but consistent. I think we stick with it for now but may need to review the solution to 4825.

I think the "proper" solution would be to only clear the RHS if the report/contact isn't available in the LHS. But that may be a bit weird given the search can be slow, so does the RHS just sit there and then disappear a few seconds later, or do we deselect it then reselect it?

I could raise an issue to get design input?

dianabarsan replied:
I agree with your idea of the proper way! let's get design input!

garethbowen replied:
Raised: https://github.com/medic/cht-core/issues/6367

garethbowen commented:
@dianabarsan I've fixed most of the bugs you found, and replied to the rest. I also added a feature that Michael was keen on which is allowing forms to specify a key to show in the live list instead of just the patient/sender's name. I'm not really happy with it because it's not as generic as I'd like but let me know what you think!

garethbowen commented:
Issue raised for nested fields bug: https://github.com/medic/cht-core/issues/6368

garethbowen commented:
Re-review please @dianabarsan !

There were a few hidden requirements (now [documented in the issue](https://github.com/medic/cht-core/issues/6291#issuecomment-620914115)) that meant I changed my approach to providing a new transition to assign a place_uuid when a case_id is generated or provided in a report.

If you're happy with this approach I'll update the documentation PR with the new configuration options.

Thanks again!

garethbowen commented on shared-libs/transitions/src/transitions/case_reports.js:
```diff
@@ -0,0 +1,116 @@
+// NB: this is very similar to a function in the registration transition, except
+//     they also allow for an empty event_type
+const messageRelevant = (msg, doc) => {
+  if (msg.event_type === 'report_accepted') {
```
Done.

garethbowen commented on shared-libs/transitions/src/transitions/case_reports.js:
```diff
@@ -0,0 +1,116 @@
+};
+
+const addMessagesToDoc = (doc, config, registrations) => {
+  if (config.messages) {
```
Done.

garethbowen commented on webapp/src/js/services/contact-view-model-generator.js:
```diff
@@ -247,6 +247,10 @@ angular.module('inboxServices').factory('ContactViewModelGenerator',
     const getHeading = function(report) {
+      const caseId = report.case_id || (report.fields && report.fields.case_id);
+      if (caseId) {
+        return false;
```
The thinking was that if it has a case then we don't care about the subject. I realised looking at this again that it was a little strange given that elsewhere we use the form's subject_key. It also helped me find a bug where the title wouldn't render correctly in the contacts content. I've changed the implementation to use the subject_key instead.

garethbowen commented:
Another review please @dianabarsan . The last commit alone should be enough. It has the change we discussed but also passes more reports to nools which was required to resolve the task when the home place is dirty.

garethbowen commented:
After failing on flakey unit tests around 10 times I've given up. I'll try again tomorrow...

dianabarsan commented on webapp/src/js/services/rules-engine.js:
```diff
@@ -128,11 +128,10 @@ angular.module('inboxServices').factory('RulesEngine', function(
-
         return RulesEngineCore
-          .updateEmissionsFor(subjectId)
+          .updateEmissionsFor(subjectIds)
```
Sorry, I was completely wrong and this isn't enough. We have to do something similar when we recalculate, after we get the reports from the db. :( Sorry

dianabarsan commented on shared-libs/registration-utils/src/index.js:
```diff
@@ -41,32 +43,46 @@ exports.isValidRegistration = (doc, settings) => {
+    }
+    if (report.fields && report.fields[prop]) {
+      return report.fields[prop];
+    }
```
Actually I was completely wrong about this, sorry!

garethbowen commented on shared-libs/rules-engine/src/pouchdb-provider.js:
```diff
@@ -8,9 +8,14 @@
-const docsOf = query => query.then(result => result.rows.map(row => row.doc).filter(existing => existing));
+const docsOf = query => {
+  return query.then(result => {
+    return uniq(result.rows.map(row => row.doc).filter(existing => existing));
```
Absolutely. Good spotting. Done.

garethbowen commented on shared-libs/rules-engine/test/rules-emitter.spec.js:
```diff
@@ -182,7 +182,7 @@ describe('rules-emitter', () => {
 
       const { tasks, targets } = await rulesEmitter.getEmissionsFor([chtDocs.contact], [chtDocs.pregnancyReport]);
-      expect(tasks.length).to.eq(1);
+      expect(tasks.length).to.eq(2);
```
Yeah. Resolved by your PR.
